File: optimizers.js
import { config, print, debug } from './config.js'
import * as tf from '@tensorflow/tfjs';

class SinglePassOptimizer {
    constructor(gameInitializer, policy, scorer) {
        this.gameInitializer = gameInitializer;
        this.policy = policy
        this.encoder = this.policy.encoder;
        this.scorer = scorer;
    }

    optimize = async function (epochs = 100) {
        var victories = 0
        var history = []
        var scores = []
        for (var x = 0; x < epochs; x++) {
            this.policy.onMatchStart()
            var game = this.gameInitializer()
            var isVictory = game.start()
            victories += (isVictory ? 1 : 0)
            var score = this.scorer.scoreGame(isVictory, game.world)
            print('Score:', score)
            scores.push(score)
            await this.policy.update(score)
            history.push(100 * victories / (x + 1))
        }
        return scores
    }
}

export class BatchPassPickBestOptimizer {
    constructor(gameInitializer, policy, scorer, batchSize=20) {
        this.gameInitializer = gameInitializer;
        this.policy = policy
        this.encoder = this.policy.encoder;
        this.scorer = scorer;
        this.batchSize = batchSize;
    }

    optimize = async function (epochs = 100) {
        var victories = 0
        var history = []
        var scores = []
        for (var x = 0; x < epochs; x++) {
            debug('Starting epoch:', x)
            var bestBatchScore = -Infinity
            var batchPickedActions = []
            for (var i = 0; i < this.batchSize; i++) {
                this.policy.onMatchStart()
                var game = this.gameInitializer()
                var isVictory = game.start()
                victories += (isVictory ? 1 : 0)
                var score = this.scorer.scoreGame(isVictory, game.world)
                if (score > bestBatchScore) {
                    bestBatchScore = score
                    batchPickedActions = this.policy.pickedActions;
                }
            }
            print('Score:', bestBatchScore)
            scores.push(bestBatchScore)
            await this.policy.update(bestBatchScore)
            history.push(100 * victories / (x + 1))
        }
        return scores
    }
}

class BatchPassOptimizer {
    constructor(game, batches) {
        this.game = game
        this.policy = game.ruleset.policy
        this.encoder = this.policy.encoder;
        this.batches = batches
    }

    optimize = function (game, epochs = 100) {
        var history = []
        var totalVictories = 0
        for (var x = 0; x < epochs; x++) {
            var victories = 0;
            var batchUpdates = [];
            for (var pass = 0; pass < this.batches; pass++) {
                this.policy.onMatchStart()
                var isVictory = this.game.start()
                victories += (isVictory ? 1 : 0)
                var score = this.scoreGame(isVictory, this.game.world) / this.policy.cardIndecesPicked.length
                this.policy.cardIndecesPicked.forEach((index) => {
                    var oneHotCard = this.encoder.getOneHotEncodedCardByIndex(index)
                    var updates = oneHotCard.map(n => n *= score)
                    batchUpdates.push(updates)
                })
            }
            batchUpdates.forEach(update => {
                this.policy.updatePolicyMatrix(tf.tensor1d(update, 'float32'))
            })
            // console.log(policy.policyTensor.arraySync())
            history.push(100 * victories / this.batches)
            totalVictories += victories
        }
        console.log(totalVictories)
        return history;
    }
}

export { SinglePassOptimizer, BatchPassOptimizer }

File: python/optimizers.py
import tensorflow as tf
from config import log
import numpy as np

class SinglePassOptimizer:
    def __init__(self, gameInitializer, policy, scorer):
        self.gameInitializer = gameInitializer
        self.policy = policy
        self.scorer = scorer

    def optimize(self, epochs=100):
        victories = 0
        history = []
        scores = []
        for x in range(epochs):
            self.policy.onMatchStart()
            match = self.gameInitializer()
            isVictory = match.start()
            victories += (1 if isVictory else 0)
            score = self.scorer.scoreGame(isVictory, match.world)
            log('Score:', score)
            scores.append(score)
            self.policy.update(score)
            history.append(100 * victories / (x + 1))
        return scores

class BatchPassPickBestOptimizer:
    def __init__(self, initializer, policy, scorer, batchSize=20):
        self.initializer = initializer
        self.policy = policy
        # self.encoder = self.policy.encoder
        self.scorer = scorer
        self.batchSize = batchSize

    def optimize(self, epochs=100):
        victories = 0
        history = []
        scores = []
        for x in range(epochs):
            log('Starting epoch:', x)
            bestBatchScore = -float('inf')
            best_picked_states = []
            best_not_picked_states = []
            for i in range(self.batchSize):
                self.policy.onMatchStart()
                match = self.initializer()
                isVictory = match.start()
                victories += (1 if isVictory else 0)
                score = self.scorer.scoreGame(isVictory, match.world)
                if score > bestBatchScore:
                    bestBatchScore = score
                    best_picked_states = [self.policy.picked_states]
                    best_not_picked_states = [self.policy.not_picked_states]
            log('Score:', bestBatchScore)
            self.policy.batch_update(best_picked_states, best_not_picked_states, [bestBatchScore])
            scores.append(bestBatchScore)
            # self.policy.update(bestBatchScore)
            history.append(100 * victories / (x + 1))
        return scores

class BatchPassOptimizer:
    def __init__(self, initializer, policy, scorer, batches):
        self.policy = policy
        self.batches = batches
        self.initializer = initializer
        self.scorer = scorer

    def optimize(self, epochs=100):
        history = []
        totalVictories = 0
        all_scores = []
        for x in range(epochs):
            victories = 0
            batch_picked = []
            batch_not_picked = []
            scores = []
            for _pass in range(self.batches):
                self.policy.onMatchStart()
                match = self.initializer()
                isVictory = match.start()
                victories += (1 if isVictory else 0)
                score = self.scorer.scoreGame(isVictory, match.world)
                scores.append(score)
                batch_picked.append(self.policy.picked_states)
                batch_not_picked.append(self.policy.not_picked_states)
            self.policy.batch_update(batch_picked, batch_not_picked, scores)
            history.append(100 * victories / self.batches)
            totalVictories += victories
            # add all scores from scores into all_scores
            all_scores.extend(scores)
        log(totalVictories)
        return all_scores





